{% extends "layout.html" %}
{% block content %}
  
<div class="container">
  <div class="navbar-default sidebar" role="navigation">
    <div class="sidebar-nav navbar-collapse">
      <ul class="nav" id="side-menu">
        <li><h5>Navigation Panel</h5></li>
        <li><a href="#expSummary"><i class="fa fa-comment-o fa-fw"></i> Summary</a></li>
        <li><a href="#expDescr"><i class="fa fa-info" aria-hidden="true" style="margin-left: 8px; padding-right:5px"></i> Description</a></li>
        <li><a href="#expGraph"><i class="fa fa-pie-chart fa-fw"></i> Graph</a></li>
        <li><a href="#expBreakdown"><i class="fa fa-table fa-fw"></i> Experiment breakdown</a></li>
        <li><a href="#tgeBreakdown"><i class="fa fa-table fa-fw"></i> TGE breakdown</a></li>
      </ul>
    </div>
  </div>

  <div id="page-wrapper">
    <div class="row">
      <div class="col-lg-12">
        <h1 class="page-header">Experiment: {{ summary.accession }}</h1>
      </div>
    </div>

    <div class="row">
    	<div class="panel panel-default" id="expSummary">
        <div class="panel-heading">
          <i class="fa fa-comment-o fa-fw"></i> Summary
          <a class="btn pull-right popoverData" data-content="Summary of information derived from this experiment." rel="popover" data-placement="left" data-original-title="Summary Info" data-trigger="hover"><i class="fa fa-info-circle pull-right helpTip"></i></a>
        </div>
        
        <div class="panel-body">
          <div class="row" style="margin-left:5px">
            <!-- <label>Data submitted by the user:</label> {{ summary.user }}</br></br> -->
            <label style="margin-top: 5px">TGEs <a class="popoverData" data-content="Total number of unique translated genomic elements observed in this experiment (some may have been identified more than once)" rel="popover" data-placement="bottom" data-original-title="Translated genomic element (TGE)" data-trigger="hover"><i class="fa fa-info-circle helpTipInner"></i></a>:</label> {{ summary.tgeNum }}</br>
            <label>Samples <a class="popoverData" data-content="Number of individual sample in this experiment." rel="popover" data-placement="bottom" data-original-title="Samples" data-trigger="hover"><i class="fa fa-info-circle helpTipInner"></i></a>:</label> {{ summary.sampleNum }}</br>
            <label>Transcripts <a class="popoverData" data-content="Number of unique transcripts assembled for RNA-seq data collected in this experiment." rel="popover" data-placement="bottom" data-original-title="Transcripts" data-trigger="hover"><i class="fa fa-info-circle helpTipInner"></i></a>:</label> {{ summary.trnNum }}<br/>
            <label>Peptides <a class="popoverData" data-content="Number of peptides identified in this experiment using mass spectrometry." rel="popover" data-placement="bottom" data-original-title="Peptides" data-trigger="hover"><i class="fa fa-info-circle helpTipInner"></i></a>:</label> {{ summary.peptUniq }} 
        	</div>
        </div>
      </div>
      <div class="panel panel-default" id="expSummary">
        <div class="panel-heading">
          <i class="fa fa-info" style="margin-left: 8px; padding-right:5px"></i> Description
          <a class="btn pull-right popoverData" data-content="A brief summary of the experimental conditions. For more detailed information, see the associated publication." rel="popover" data-placement="left" data-original-title="Experiment Description" data-trigger="hover"><i class="fa fa-info-circle pull-right helpTip"></i></a>
        </div>
        
        <div class="panel-body">
          <div class="row" style="margin-left:5px">
            {% if summary.accession == 'PIT000001' %}
              <p>Data from a human ovarian cancer experiment. The whole transcriptome was sequenced for this experiment. Mass-spectrometry was carried out only on the extra cellular matrix. Tissues from different stage, grade have been used for this experiment to find underlying mechanism of drug response.</p><br/>
              <p>Funded by: European Research Council (ERC322566); the Cancer Research UK (A16354)</p><br/>
              <p style='font-size: 11px; font-style: italic;'>* Deconstructing a Metastatic Human Tumor Microenvironment. O.M.T. Pearce, R. Delaine-Smith, E. Maniati, S. Nichols, J. Wang, S. Böhm, V. Rajeeve, D. Ullah, P. Chakravarty, R.R. Jones, A. Monfort, T. Dowe, J. Connelly, J.D. Brenton, C. Chelala, P. R. Cutillas, M. Lockley, C. Bessant, M. Knight, F.R. Balkwill (submitted to Nature)</p>
            {% elif summary.accession == 'PIT000002' %}
              <p>Data from the immortalised <span style='font-style:italic'>Aedes Aegypti </span>(mosquito) cell line Aag2, commonly used for arbovirus research.</p><br/>
              <p>Funded by: Medical Research Council (G0801973); Sir Henry Wellcome Fellowship (096062)</p><br/>
              <p style='font-size: 11px; font-style: italic;'>* Proteomics Informed by Transcriptomics for Characterising Active Transposable Elements and Genome Annotation in Aedes Aegypti . K. Maringer, A. Yousuf, K.J. Heesom, J. Fan, A. Fernandez-Sesma, C. Bessant, D.A. Matthews, A.D. Davidson (submitted to BMC Biology)</p>
            {% elif summary.accession == 'PIT000003' %}
              <p>This experiment was carried out on Human HeLa cell line infected with adenovirus.</p><br/>
              <p>Funded by: Wellcome Trust (083604), BBSRC (BB/I00095X/1).</p><br/>
              <p style='padding-right: 15px'>De novo derivation of proteomes from transcriptomes for transcript and protein identification. V.C. Evans, G. Barker, K.J. Heesom, J. Fan, C. Bessant, D.A. Matthews. Nature Methods, 2012. 9(12): p. 1207-11.</p>
            {% elif summary.accession == 'PIT000004' %}
              <p>Data from <span style='font-style:italic'>Pteropus alecto</span> kidney cell line (PaKiT03) exposed to Nelson Bay orthoreovirus (NBV). <span style='font-style:italic'>Pteropus alecto</span> is an Australian bat, also known a the black flying fox.</p><br/>
              <p>Funded by: BBSRC (BB/M02542X/1); Australian Research Council (CE11001007); PhD fellowship from the University of Melbourne</p><br/>
              <p style='font-size: 11px; font-style: italic;'>* Cell tropism of Nelson Bay Orthoreovirus is Influenced by the Type 1 I IFN Signalling Pathway.  L. Mok, J.W. Wynne, M. Tachedjian, B. Shiell, K. Ford , D.A. Matthews, A. Bacic, W.P. Michalskia (submitted to Journal of Virology)</p>
            {% elif summary.accession == 'PIT000005' %}
              <p>Data from mouse fibroblast (L929) cells exposed to Nelson Bay orthoreovirus (NBV).</p><br/>
              <p>Funded by: BBSRC (BB/M02542X/1); Australian Research Council (CE11001007); University of Melbourne; CSIRO.</p><br/>
              <p style='font-size: 11px; font-style: italic;'>* Cell tropism of Nelson Bay Orthoreovirus is Influenced by the Type 1 I IFN Signalling Pathway.  L. Mok, J.W. Wynne, M. Tachedjian, B. Shiell, K. Ford , D.A. Matthews, A. Bacic, W.P. Michalskia (submitted to Journal of Virology)</p>
            {% endif %}
          </div>
        </div>
      </div>

      <div id="expGraph" class="panel panel-default sunburst" style="position: relative !important;">
        <div class="panel-heading">
          <i class="fa fa-pie-chart fa-fw"></i> Graph
          <a class="btn pull-right popoverData" data-content="Visual summary of the Translated Genomic Elements (TGEs) identified, together with their type and information about their supporting peptides. Move the mouse cursor over the plot for quantitative information." rel="popover" data-placement="left" data-original-title="Graph Info" data-trigger="hover"><i class="fa fa-info-circle pull-right helpTip"></i></a>
        </div>
        
        <div class="col-md-8 col-md-offset-2">
          <div id="sequence"></div>
          <div id="chart">
            <div id="explanation">
              <span id="percentage"><i class="fa fa-spinner fa-spin fa-3x fa-fw"></i><span class="sr-only">Loading...</span></span>
            </div>
          </div>
        </div>
        <div style="clear: both;"></div>
      </div>

      <div class="panel panel-default" id="expBreakdown">
        <div class="panel-heading">
            <i class="fa fa-table fa-fw"></i> Experiment breakdown
            <a class="btn pull-right popoverData" data-content="Information about the samples analysed within the experiment." rel="popover" data-placement="left" data-original-title="Experiment Breakdown" data-trigger="hover"><i class="fa fa-info-circle pull-right helpTip"></i></a>
        </div>
        
        <div class="panel-body">
            <table id = "smplTable" class="table table-striped table-bordered dataTable no-footer" >
            <thead>
              <tr>
                <!-- <th>Sample ID&nbsp;<a class="popoverData" data-content="The accession number of a sample in the PITDB database." rel="popover" data-placement="bottom" data-original-title="Sample ID" data-trigger="hover"><i class="fa fa-info-circle"></i></a></th> -->
                <th>Sample Name&nbsp;<a class="popoverData" data-content="Detailed sample name" rel="popover" data-placement="bottom" data-original-title="Sample Name" data-trigger="hover"><i class="fa fa-info-circle"></i></a></th>
                <th>Instances of TGEs&nbsp;<a class="popoverData" data-content="Number of TGEs observed in each sample." rel="popover" data-placement="bottom" data-original-title="Instances of TGEs" data-trigger="hover"><i class="fa fa-info-circle"></i></a></th>
                <th>Unique Peptide Count&nbsp;<a class="popoverData" data-content="Number of unique peptides identified in this sample using mass spectrometry." rel="popover" data-placement="bottom" data-original-title="Peptide Count" data-trigger="hover"><i class="fa fa-info-circle"></i></a></th>
              </tr>
            </thead>
            <tbody>
              {% for sample in sampleList %}
                <tr>
                  <!-- <td>{{ sample.id }}</td> -->
                  <td>{{ sample.name }}</td>
                  <td>{{ sample.tgeNum }}</td>
                  <td>{{ sample.pepNum }}</td>
                </tr>
              {% endfor %}
            </tbody>
          </table>
        </div>
      </div>


      <div class="panel panel-default" id="tgeBreakdown">
        <div class="panel-heading">
          <i class="fa fa-table fa-fw"></i> Translated Genomic Elements (TGEs)
          <a class="btn pull-right popoverData" data-content="List of experimentally TGEs that have been observed in this experiment." rel="popover" data-placement="left" data-original-title="Translated genomic elements (TGEs)" data-trigger="hover"><i class="fa fa-info-circle pull-right helpTip"></i></a>
        </div>
        
        <div class="panel-body">
          <table class="table table-striped table-bordered dataTable no-footer" id="exptgeTable" style="display: none;">
          <thead>
            <tr>
              <th>TGE Accession&nbsp;<a class="popoverData" data-content="Unique identifier of a TGE in the PITDB database." rel="popover" data-placement="bottom" data-original-title="TGE Accession" data-trigger="hover"><i class="fa fa-info-circle"></i></a></th>
              <th>TGE Type(s)&nbsp;<a class="popoverData" data-content="One TGE may have multiple types, as it may have a different status in different species. Use the table’s search and sort capabilities to look for specific TGEs or TGE types." rel="popover" data-placement="bottom" data-original-title="TGE Type(s)" data-trigger="hover"><i class="fa fa-info-circle"></i></a></th>
              <th>Uniprot ID(s)&nbsp;<a class="popoverData" data-content="Accession number of homologous Uniprot protein (if there is one). Click the accession number to see what else the database knows about this protein." rel="popover" data-placement="bottom" data-original-title="Uniprot ID(s)" data-trigger="hover"><i class="fa fa-info-circle"></i></a></th>
              <th>Organism(s)&nbsp;<a class="popoverData" data-content="The organism(s) in which the TGE has been observed." rel="popover" data-placement="bottom" data-original-title="Organism(s)" data-trigger="hover"><i class="fa fa-info-circle helpTipInner"></i></a>:</th>
            </tr>
          </thead>
          <tbody>
            {% for tge in tges %}
              <tr>
                <td>
                <a href="tge?accession={{ tge.accession }}">{{ tge.accession }}</a>
                </td>
                <td>{{ tge.tge_class }}</td>
                <td>
                  {% for uniprotID in tge.uniprot_id.split(',') %}
                    {% if uniprotID != '-' %}
                      <a href="protein?uniprot={{ uniprotID }}">{{ uniprotID }}</a>
                    {% else %}
                      <span style="font-size:18px">{{ uniprotID }}</span>
                    {% endif %}{% if not loop.last %},{% endif %}
                  {% endfor %}</br>
                </td>
                <td>
                  {% for organism in tge.organisms.split(',') %}
                  {% if organism != '-' %}
                    <a href="organism?organism={{ organism }}">{{ organism }}</a>
                  {% else %}
                    <span style="font-size:18px">{{ organism }}</span>
                  {% endif %}{% if not loop.last %},{% endif %}
                {% endfor %}</br>
                </td>
              </tr>
            {% endfor %}
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>

{% endblock %}