{% extends "layoutPlots.html" %}
{% block content %}
  
<div class="navbar-default sidebar" role="navigation">
  <div class="sidebar-nav navbar-collapse">
    <ul class="nav" id="side-menu">
      <li><h5>Navigation Panel</h5></li>
      <li><a href="#protGraph"><i class="fa fa-pie-chart fa-fw"></i> Graph</a></li>
      <li><a href="#protSummary"><i class="fa fa-comment-o fa-fw"></i> TGEs</a></li>
      <li><a href="#genoGraph"><i class="fa fa fa-picture-o fa-fw"></i> Genoverse</a></li>
    </ul>
    </div>
  </div>

  <div id="page-wrapper">
    <div class="row">
      <div class="col-lg-12">
        <h1 class="page-header">Uniprot ID: <a href="http://www.uniprot.org/uniprot/{{ uniprot }}" target="_blank">{{ uniprot }}</a></h1>
      </div>
    </div>

    <div class="row">
      <div id="protGraph" class="panel panel-default sunburst" style="position: relative !important;">
        <div class="panel-heading">
          <i class="fa fa-pie-chart fa-fw"></i> Graph
        </div>

        <span><p><h5 style="margin-left: 15px; padding-top: 10px">Graph of organism breakdown</h5></p></span>
        <div class="col-md-8 col-md-offset-2">
          <div id="sequence"></div>
          <div id="chart">
            <div id="explanation">
              <span id="percentage"><i class="fa fa-spinner fa-spin fa-3x fa-fw"></i><span class="sr-only">Loading...</span></span>
            </div>
          </div>
        </div>
        <div style="clear: both;"></div>
      </div>

    	<div class="panel panel-default" id="protSummary">
        <div class="panel-heading">
          <i class="fa fa-comment-o fa-fw"></i> TGEs
        </div>
        
        <div class="panel-body">
          <div class="row" style="margin-left:5px">
            <table id="proteinRes" class="table table-striped table-bordered dataTable no-footer" class="searchRes">
            <thead>
              <tr>
                <th>TGE Accession</th>
                <th>TGE Type</th>
              </tr>
            </thead>
            <tbody>
              {% for tge in tges %}
                <tr>
                  <td><a href="tge?accession={{ tge.accession }}" target="_blank">{{ tge.accession }}</a></td>
                  <td>{{ tge.tge_class }}</td>
                </tr>
              {% endfor %}
            </tbody>
          </table>
        </div>
      </div>
    </div>

    <div>
      <div class="panel panel-default" id="genoGraph">
        <div class="panel-heading">
          <i class="fa fa fa-picture-o fa-fw"></i> Genoverse
        </div>
        
        <div class="panel-body">
          <div id="genoverse"></div>
        </div>
    </div>
  </div>
</div>

<script src="/static/js/ext/jquery-2.1.4.min.js"></script>

<script type="text/javascript" src="/static/js/genoverse.combined.js"></script>

<script>
  var chrParam = getParam("seqid");
  
  if (chrParam == null || chrParam == '') {
    chrParam = 2;
  }
  var organism = "{{ genoverse.organism }}";
  var genomeName, org; 

  switch (organism) {
    case 'Homo sapiens':
      genomeName = 'grch38';
      org = 'human';
      break;
    case 'Aedes aegypti':
      genomeName = 'grch38';
      break;
    case 'Pteropus alecto':
      genomeName = 'grch38';
      break;
    case 'Mus musculus':
      genomeName = 'mouse';
      org = 'mouse';
      break;
}

  var genoverseConfig = {
    container : '#genoverse',
    width     : 900,
    genome    : genomeName,
    chr       : "{{ genoverse.chr }}",
    start     : "{{ genoverse.start }}",
    end       : "{{ genoverse.end }}",
    plugins   : [ 'controlPanel', 'karyotype', 'trackControls', 'resizer' ],
    tracks    : [
      Genoverse.Track.Scalebar,
      Genoverse.Track.extend({
        name  : 'Sequence',
        model : Genoverse.Track.Model.Sequence.Ensembl.extend({
          url : '//rest.ensembl.org/sequence/region/'+org+'/__CHR__:__START__..__END__:0?content-type=application/json'
        }),
        view      : Genoverse.Track.View.Sequence,
        resizable : 'auto',
        100000    : false
      }),
      Genoverse.Track.extend({
        name   : 'Genes',
        height : 200,
        info   : 'Ensembl API genes & transcripts, see <a href="http://beta.rest.ensembl.org/" target="_blank">beta.rest.ensembl.org</a> for more details',

        // Different settings for different zoom level
        2000000: { // This one applies when > 2M base-pairs per screen
          labels : false
        },
        100000: { // more than 100K but less then 2M
          labels : true,
          model  : Genoverse.Track.Model.Gene.Ensembl.extend({
            url : '//rest.ensembl.org/overlap/region/'+org+'/__CHR__:__START__-__END__?feature=gene;content-type=application/json'
          }),
          view   : Genoverse.Track.View.Gene.Ensembl
        },
        1: { // > 1 base-pair, but less then 100K
          labels : true,
          model  : Genoverse.Track.Model.Transcript.Ensembl.extend({
              url : '//rest.ensembl.org/overlap/region/'+org+'/__CHR__:__START__-__END__?feature=transcript;feature=exon;feature=cds;content-type=application/json'
          }),
          view   : Genoverse.Track.View.Transcript.Ensembl
        }
      }),
      Genoverse.Track.extend({
        name   : 'GFF3',
        height : 400,
        //info   : 'Ensembl API genes & transcripts, see <a href="http://beta.rest.ensembl.org/" target="_blank">beta.rest.ensembl.org</a> for more details',
        url : '/genoverse/'+"{{ uniprot }}",
        //info   : 'Ensembl API genes & transcripts, see <a href="http://beta.rest.ensembl.org/" target="_blank">beta.rest.ensembl.org</a> for more details',
        view   : Genoverse.Track.View.Transcript.extend({
                setFeatureColor : function () {}
        }),    
        model : Genoverse.Track.Model.Transcript.extend(
        {
        dataType : 'text',
            typeMap : {
                exon  : 'exon',
                cds   : 'cds',
                mrna  : 'mrna',
                gene  : 'gene'
            },
        parseData: function (text) {

      //alert(text)
      //alert("1. parent:"+feature.parent)
      //alert("1. type:"+feature.type)
      var lines = text.split('\n');
                var gene;
          for (var i = 0; i < lines.length; i++) {
                    if (!lines[i].length || lines[i].indexOf('#') === 0) {
                        continue;
                    }
                    //alert(lines[i])
                    var fields = lines[i].split('\t');
            
                    if (fields.length < 5) {
                        continue;
                    }
                  
                    if (fields[0] === this.browser.chr || fields[0].toLowerCase() === 'chr' + this.browser.chr || fields[0].match('[^1-9]' + this.browser.chr + '$'))
                    {
                        var feature = {};
    
                        feature.id     = fields.slice(0, 5).join('|');
                        feature.start  = parseInt(fields[3], 10);
                        feature.end    = parseInt(fields[4], 10);
                        feature.source = fields[1];
                        feature.type   = fields[2];
                        feature.score  = fields[5];
                        feature.strand = fields[6] + '1';
    
                        if (fields[8]) {
                            var frame = fields[8].split(';');
                            
                            for (var j = 0; j < frame.length; j++) {
                                var keyValue = frame[j].split('=');
                                
                                if (keyValue.length === 2) {
                                    feature[keyValue[0].toLowerCase()] = keyValue[1];
                                }
                            }
                        }
                        //alert("0. parent:"+feature.parent)
                        //alert("0. type:"+feature.type)
                        //alert(gene)
                        //alert("type:"+feature.type.toLowerCase())
                        //alert("type:"+this.typeMap.mrna.toLowerCase())
                        // sub-feature came earlier than parent feature
                        //if (feature.parent && feature.type.toLowerCase() === this.typeMap.mrna.toLowerCase() && !this.featuresById[feature.id]) {
                        if(feature.type.toLowerCase() === this.typeMap.gene.toLowerCase())
                        {
                            //alert("gene");
                            if(typeof gene !== "undefined" && gene)
                            {
                                //alert("gene features entering");
                                for(var g=0; g<gene.mrna.length; g++)
                                {
                                    this.insertFeature(gene.mrna[g]);
                                }
                            }
                            feature.mrna=[];
                            gene=feature;
                            //alert("gene id:"+gene.id);
                            //alert("gene mRNA:"+gene.mrna.length);
                        }
                        else if(feature.type.toLowerCase() === this.typeMap.mrna.toLowerCase())
                        {
                            //alert('mRNA');
                            feature.exons=[];
                            feature.cds=[];
                            if(gene!==undefined)
                            {
                                gene.mrna.push(feature);
                                //alert("gene not undefined:"+gene.mrna.length);
                            }
                            //alert('mRNA2');
                        }
                        else if (feature.parent && feature.type.toLowerCase() === this.typeMap.exon.toLowerCase()) {
                            //alert("exon");
                            var idx = gene.mrna.length - 1;
                            
                            //alert("idx="+idx)
                            gene.mrna[idx].exons.push(feature); //assuming last mRNA is the parent of this exon
                            //alert("exon pushed");
                        }
                        else if (feature.parent && feature.type.toLowerCase() === this.typeMap.cds.toLowerCase()) {
                            //alert("cds");
                            var idx=gene.mrna.length-1;
                            feature.color = '#B0171F';
                            gene.mrna[idx].cds.push(feature); //assuming last mRNA is the parent of this cds
                            //alert("cds pushed");
                        }
                        
                    } //end of if
          }//end of forloop
                if(typeof gene !== "undefined" && gene)
                {
                    //alert("gene features entering");
                    for(var g=0; g<gene.mrna.length; g++)
                    {
                        this.insertFeature(gene.mrna[g]);
                    }
                }
                //alert()
            } //end of function
      })
        })//,
    ]
  };
  
  document.addEventListener('DOMContentLoaded', function () { 
    window.genoverse = new Genoverse(genoverseConfig); 
  });

  function getParam(name){
    if(name=(new RegExp('[?&]'+encodeURIComponent(name)+'=([^&]*)')).exec(location.search))
      return decodeURIComponent(name[1]);
  }
</script>

{% endblock %}
