{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import psycopg2\n",
    "import yaml\n",
    "import sys\n",
    "import json\n",
    "import csv\n",
    "import re\n",
    "\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the contents of the yaml file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "stream = open(\"db.yml\", \"r\")\n",
    "db = yaml.load(stream)\n",
    "stream.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Print the experiment title\n",
    "\n",
    "# Print the experiment title\n",
    "\n",
    "expName  = \"Mosquito\"\n",
    "dataPath = \"Data/\"+ expName\n",
    "print (dataPath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Insert into the *experiment* table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "\n",
    "    expINS = \"INSERT INTO experiment (title) VALUES ('Mosquito') \" \n",
    "    cur.execute(expINS,)\n",
    "    con.commit()\n",
    "        \n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Insert into the *sample* table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "    \n",
    "    # We cannot have a duplicate name and exp_id ( AND exp_id = %)\n",
    "    smpINS = 'INSERT INTO sample (name, exp_id) SELECT %s, %s WHERE NOT EXISTS (SELECT id FROM sample WHERE name = %s AND exp_id = %s);' \n",
    "    \n",
    "    path = dataPath + \"/AminoAcids-or-ORFs-orTGEs\"\n",
    "    \n",
    "    for i in os.listdir(path):\n",
    "        if i.endswith(\".assemblies.fasta.transdecoder.pep.identified.fasta\"): \n",
    "            print(i)\n",
    "            sampleName = str(i[0:i.find('.')])\n",
    "            \n",
    "            query = (\"SELECT id FROM experiment WHERE title = 'Mosquito'\")\n",
    "            cur.execute(query)\n",
    "            expID = cur.fetchone()[0]\n",
    "            \n",
    "            cur.execute(smpINS, (sampleName, expID, sampleName, expID))\n",
    "            con.commit()\n",
    "        \n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the data into the *tge* table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Load the data into the <tge> table\n",
    "\n",
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "    \n",
    "    # We cannot have duplicate TGE amino acid sequences \n",
    "    tgeINS = 'INSERT INTO tge (amino_seq) SELECT %s WHERE NOT EXISTS (SELECT id FROM tge WHERE amino_seq = %s);' \n",
    "    \n",
    "    tgePath = dataPath + \"/AminoAcids-or-ORFs-orTGEs/\"\n",
    "        \n",
    "    for i in os.listdir(tgePath):\n",
    "        if i.endswith(\".assemblies.fasta.transdecoder.pep.identified.fasta\"): \n",
    "            print(i)\n",
    "             \n",
    "            for input in SeqIO.parse(tgePath+i, \"fasta\"):\n",
    "                tgeSeq = input.seq\n",
    "                tgeSeq = str(tgeSeq[0:tgeSeq.find('*')])\n",
    "                cur.execute(tgeINS, (tgeSeq, tgeSeq))\n",
    "                con.commit()\n",
    "\n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Load the data into the <observation> table\n",
    "\n",
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "    \n",
    "    tgeINS = 'INSERT INTO observation (tge_id, sample_id, name, description) SELECT %s , %s , %s, %s WHERE NOT EXISTS (SELECT id FROM observation WHERE description = %s AND sample_id = %s);' \n",
    "    \n",
    "    tgePath = dataPath + \"/AminoAcids-or-ORFs-orTGEs/\"\n",
    "        \n",
    "    for i in os.listdir(tgePath):\n",
    "        if i.endswith(\".assemblies.fasta.transdecoder.pep.identified.fasta\"): \n",
    "            print (i)\n",
    "            \n",
    "            sampleName = str(i[0:i.find('.')])\n",
    "            # WE SHOULD TAKE THE EXPERIMENT NAME TOO \n",
    "            query = (\"SELECT id FROM sample WHERE sample.name = '\"+sampleName+\"'\")\n",
    "            cur.execute(query)\n",
    "            sampleID = cur.fetchone()[0]\n",
    "            \n",
    "            for input in SeqIO.parse(tgePath+i, \"fasta\"):\n",
    "                tgeName  = input.id\n",
    "                tgeName  = tgeName[0:tgeName.find('|')]\n",
    "                \n",
    "                tgeSeq = input.seq\n",
    "                tgeSeq = str(tgeSeq[0:tgeSeq.find('*')])\n",
    "            \n",
    "                query = (\"SELECT id FROM tge WHERE tge.amino_seq = '\"+tgeSeq+\"'\")\n",
    "                cur.execute(query)\n",
    "                tgeID = cur.fetchone()[0]\n",
    "            \n",
    "                cur.execute(tgeINS, (tgeID, sampleID, tgeName, input.id, input.id, sampleID))\n",
    "                con.commit()\n",
    "                \n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Load the data into the <transcript> table\n",
    "\n",
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "\n",
    "    trsINS = 'INSERT INTO transcript (obs_id, dna_seq) SELECT %s ,%s WHERE NOT EXISTS (SELECT id FROM transcript WHERE obs_id = %s AND dna_seq  = %s) AND EXISTS (SELECT id FROM observation WHERE name = %s AND sample_id = %s);'\n",
    "    \n",
    "    trsPath = dataPath + \"/transcripts/\"\n",
    "    \n",
    "    for i in os.listdir(trsPath):\n",
    "        if i.endswith(\".assemblies.fasta.identified.fasta\"): \n",
    "            print (i)\n",
    "            \n",
    "            sampleName = str(i[0:i.find('.')])\n",
    "            query = (\"SELECT id FROM sample WHERE sample.name = '\"+sampleName+\"'\")\n",
    "            cur.execute(query)\n",
    "            sampleID = cur.fetchone()[0]\n",
    "            \n",
    "            for input in SeqIO.parse(trsPath+i, \"fasta\"):\n",
    "                trsName = input.id\n",
    "                trsSeq  = str(input.seq)\n",
    "                \n",
    "                query = (\"SELECT id FROM observation WHERE name = '\"+trsName+\"' AND sample_id=\"+str(sampleID))\n",
    "                cur.execute(query)\n",
    "                tgeObsID = cur.fetchone()[0]\n",
    "\n",
    "                cur.execute(trsINS, (tgeObsID, trsSeq, tgeObsID, trsSeq, trsName, sampleID))\n",
    "                con.commit()\n",
    "        \n",
    "\n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## update annotation in observation\n",
    "\n",
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "\n",
    "    tgeALT = 'UPDATE observation SET organism = %s, uniprot_id = %s, protein_name= %s, protein_descr = %s, gene_name = %s, tge_class = %s, variation = %s, long_description = %s  WHERE description = %s AND sample_id = %s'\n",
    "    \n",
    "    annPath = dataPath + \"/Annotation/\"\n",
    "    \n",
    "    for i in os.listdir(annPath):\n",
    "        if i.endswith(\".pep_details_annotation.csv\"): \n",
    "            print (i)\n",
    "            \n",
    "            sampleName = str(i[0:i.find('.')])\n",
    "            query = (\"SELECT id FROM sample WHERE sample.name = '\"+sampleName+\"'\")\n",
    "            \n",
    "            cur.execute(query)\n",
    "            sampleID = cur.fetchone()[0]\n",
    "            \n",
    "            annotation = pd.read_csv(annPath+i)\n",
    "                \n",
    "            for i in range(len(annotation)): \n",
    "                descr  = annotation['ORF Id'][i]\n",
    "                tgeDescr  = descr[0:descr.find(' ')]\n",
    "                \n",
    "                uniprotID = annotation['Protein ID'][i]\n",
    "#                 m = re.search(\"(?<=\\|).*?(?=\\|)\", protID)\n",
    "                if uniprotID is None:\n",
    "                    uniprotID = '-'\n",
    "            \n",
    "                proteinName  = str(annotation['Protein Name'][i])\n",
    "                geneName     = str(annotation['Gene Name'][i])\n",
    "                proteinDescr = str(annotation['Protein description'][i])\n",
    "                tgeClass  = str(annotation['Class'][i])\n",
    "                variation = int(annotation['Variation'][i])\n",
    "                species   = str(annotation['Species'][i])\n",
    "                \n",
    "#                 print(descr)\n",
    "#                 print(tgeDescr)\n",
    "#                 print(uniprotID)\n",
    "#                 print(proteinName)\n",
    "#                 print(geneName)\n",
    "#                 print(proteinDescr)\n",
    "#                 print(tgeClass)\n",
    "#                 print(variation)\n",
    "#                 print(species)\n",
    "                \n",
    "                cur.execute(tgeALT, (species, uniprotID, proteinName, proteinDescr, geneName, tgeClass, variation, descr, tgeDescr, sampleID))\n",
    "                con.commit()\n",
    "\n",
    "                \n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## update observation table with peptide number \n",
    "\n",
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "\n",
    "    tgeALT = 'UPDATE observation SET peptide_num = %s WHERE description = %s AND sample_id = %s'\n",
    "    \n",
    "    pepPath = dataPath + \"/PSMs-Peptides-ORFs/\"\n",
    "    \n",
    "\n",
    "    for i in os.listdir(pepPath):\n",
    "        if i.endswith(\"+fdr+th+grouping+prt_filtered.csv\"): \n",
    "            print (i)\n",
    "            \n",
    "            sampleName = str(i[0:i.find('+')])\n",
    "            query = (\"SELECT id FROM sample WHERE sample.name = '\"+sampleName+\"'\")\n",
    "            cur.execute(query)\n",
    "            sampleID = cur.fetchone()[0]\n",
    "            \n",
    "            tge = pd.read_csv(pepPath+i)\n",
    "\n",
    "            for i in range(len(tge)): \n",
    "                descr  = str(tge['description'][i])\n",
    "                tgeID  = descr[0:descr.find(' ')]\n",
    "                pepNum = int(tge['distinct peptide sequences'][i])\n",
    "                \n",
    "                cur.execute(tgeALT, (pepNum, tgeID, sampleID))\n",
    "                con.commit()\n",
    "\n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Insert into peptide \n",
    "\n",
    "\n",
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "\n",
    "    insPep = 'INSERT INTO peptide (aa_seq) SELECT %s WHERE NOT EXISTS (SELECT id FROM peptide WHERE aa_seq = %s);'\n",
    "    \n",
    "    pepPath = dataPath + \"/PSMs-Peptides-ORFs/\"\n",
    "    \n",
    "\n",
    "    for i in os.listdir(pepPath):\n",
    "        if i.endswith(\"+fdr+th+grouping_filtered.csv\"): \n",
    "            print (i)\n",
    "            \n",
    "            peptides = pd.read_csv(pepPath+i)\n",
    "\n",
    "            for i in range(len(peptides)): \n",
    "                m = re.search(\"(?<=index=).*?$\", peptides['Spectrum ID'][i])\n",
    "                specID = m.group(0)\n",
    "                \n",
    "                amino   = peptides['Sequence'][i]\n",
    "#                 calc_mz = peptides['Calc m/z'][i]\n",
    "#                 exp_mz  = peptides['Exp m/z'][i]\n",
    "                \n",
    "                cur.execute(insPep, (amino, amino))\n",
    "                con.commit()\n",
    "    \n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Peptides for each TGE\n",
    "\n",
    "\n",
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "\n",
    "    insTgePep = 'INSERT INTO tge_peptide (obs_id, peptide_id) SELECT %s , %s WHERE NOT EXISTS (SELECT id FROM tge_peptide WHERE obs_id = %s AND peptide_id  = %s);'\n",
    "    \n",
    "    pepPath = dataPath + \"/PSMs-Peptides-ORFs/\"\n",
    "    \n",
    "\n",
    "    for i in os.listdir(pepPath):\n",
    "        if i.endswith(\"+fdr+th+grouping_filtered.csv\"): \n",
    "            print (i)\n",
    "            \n",
    "            sampleName = str(i[0:i.find('+')])\n",
    "            query = (\"SELECT id FROM sample WHERE sample.name = '\"+sampleName+\"'\")\n",
    "            cur.execute(query)\n",
    "            sampleID = cur.fetchone()[0]\n",
    "            \n",
    "            peptides = pd.read_csv(pepPath+i)\n",
    "             \n",
    "            \n",
    "            for i in range(len(peptides)): \n",
    "#                 m = re.search(\"(?<=index=).*?$\", peptides[i][0])\n",
    "#                 specID = m.group(0)\n",
    "\n",
    "                amino   = peptides['Sequence'][i]\n",
    "                \n",
    "                query = (\"SELECT id FROM peptide WHERE aa_seq = '\"+amino+\"'\") \n",
    "                cur.execute(query)\n",
    "                pepID = cur.fetchone()[0]\n",
    "                \n",
    "                proteinacc = peptides['proteinacc_start_stop_pre_post_;'][i]\n",
    "                arr =  proteinacc.split(';')\n",
    "                \n",
    "                for x in range(0, len(arr)):\n",
    "                    n=2\n",
    "                    tge = arr[x]                    \n",
    "                    m=re.match(r'^((?:[^_]*_){%d}[^_]*)_(.*)' % (n-1), tge)\n",
    "                    tgeDescr = m.groups()[0]\n",
    "                    \n",
    "                    query = (\"SELECT id FROM observation WHERE description = '\"+tgeDescr+\"' AND sample_id = \"+str(sampleID))\n",
    "                    cur.execute(query)\n",
    "                    tgeID = cur.fetchone()\n",
    "                    \n",
    "                    if tgeID:\n",
    "                        tgeID = tgeID[0]\n",
    "                        \n",
    "                        cur.execute(insTgePep, (tgeID, pepID,tgeID, pepID ))\n",
    "                        con.commit()\n",
    "\n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Insert into psm \n",
    "\n",
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "\n",
    "    insSpec = 'INSERT INTO psm (peptide_id, name, spectrum_id, title, location, retention, calc_mz, exp_mz, charge, modifications, raw_score, denovo_score, spec_evalue, evalue, qvalue, pep_qvalue, local_fdr, q_value, fdr_score ) SELECT %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s WHERE NOT EXISTS (SELECT id FROM psm WHERE spectrum_id = %s);'\n",
    "    \n",
    "    pepPath = dataPath + \"/PSMs-Peptides-ORFs/\"\n",
    "    \n",
    "\n",
    "    for i in os.listdir(pepPath):\n",
    "        if i.endswith(\".assemblies.fasta.transdecoder.pep+fdr+th+grouping_filtered.csv\"): \n",
    "            print i\n",
    "            \n",
    "            with open(pepPath+i,'rb') as fin: \n",
    "                dr = csv.DictReader(fin)\n",
    "                psms = [( i['PSM_ID'], i['Spectrum ID'], i['Spectrum Title'], i['Raw data location'], \n",
    "                            i['Retention Time (s)'], i['Calc m/z'], i['Exp m/z'], i['Charge'], i['Modifications'],\n",
    "                            i['MS-GF:RawScore'], i['MS-GF:DeNovoScore'], i['MS-GF:SpecEValue'], i['MS-GF:EValue'], \n",
    "                            i['MS-GF:QValue'], i['MS-GF:PepQValue'], i['PSM-level local FDR'], i['PSM-level q-value'], \n",
    "                            i['PSM-level FDRScore'], i['Sequence']) for i in dr]\n",
    "\n",
    "            for i in range(len(psms)): \n",
    "                location  = psms[i][3]\n",
    "                \n",
    "                m = re.search(\"(?<=index=).*?$\", psms[i][1])\n",
    "                specID = m.group(0)\n",
    "                \n",
    "                specTitle = psms[i][2]\n",
    "                \n",
    "                amino   = psms[i][18]\n",
    "                \n",
    "                query = (\"SELECT id FROM peptide WHERE aa_seq = '\"+amino+\"'\") \n",
    "                cur.execute(query)\n",
    "                pepID = cur.fetchone()[0]\n",
    "                \n",
    "                cur.execute(insSpec, (pepID, psms[i][0], specID, specTitle, location, psms[i][4], psms[i][5], psms[i][6], psms[i][7], psms[i][8], psms[i][9], psms[i][10], psms[i][11], psms[i][12], psms[i][13], psms[i][14], psms[i][15], psms[i][16], psms[i][17], specID))\n",
    "                con.commit()\n",
    "    \n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "\n",
    "try:\n",
    "    con = psycopg2.connect(database=db[\"test\"][\"database\"], user=db[\"test\"][\"username\"], host=db[\"test\"][\"host\"], password=db[\"test\"][\"password\"])\n",
    "    cur = con.cursor()\n",
    "\n",
    "    insSpec = 'INSERT INTO variation (obs_id, chrom, pos, alt, qual, var_type, qpos, peptide_num, peptides) SELECT %s, %s, %s, %s, %s, %s, %s, %s, %s;'\n",
    "    \n",
    "    variationPath = dataPath + \"/Variations-proVCF/\"\n",
    "    \n",
    "\n",
    "    for filename in os.listdir(pepPath):\n",
    "        if filename.endswith(\".assemblies.fasta.transdecoder.pep_pepEvd.vcf\"): \n",
    "            print filename\n",
    "            \n",
    "            vcf_reader = vcf.Reader(open(variationPath+filename, 'r'))\n",
    "\n",
    "            for record in vcf_reader:\n",
    "                record = next(vcf_reader)\n",
    "                print record.CHROM\n",
    "                print record.ALT\n",
    "                print record.QUAL\n",
    "                print record.INFO['SubjectId']\n",
    "                print record.INFO['Alignment'][0]\n",
    "                print record.INFO['Type']\n",
    "                print record.INFO['QPOS']\n",
    "                print record.INFO['PeptideCount']\n",
    "                print record.INFO['UniquePeptideCount']\n",
    "                print record.INFO['Score']\n",
    "            \n",
    "except (psycopg2.DatabaseError, e):\n",
    "    if con:\n",
    "        con.rollback()\n",
    "    \n",
    "    print('Error %s' % e)    \n",
    "    sys.exit(1)\n",
    "    \n",
    "finally:\n",
    "    if con:\n",
    "        con.close()\n",
    "        print(\"Connection closed\")\"\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
